notes.lua

```diff
@@ -0,0 +1,324 @@
+--@ module = true
+
+local gui = require('gui')
+local widgets = require('gui.widgets')
+local textures = require('gui.textures')
+local overlay = require('plugins.overlay')
+local guidm = require('gui.dwarfmode')
+local text_editor = reqscript('internal/journal/text_editor')
+
+local green_pin = dfhack.textures.loadTileset('hack/data/art/note-green-pin.png', 32, 32, true)
+
+NotesOverlay = defclass(NotesOverlay, overlay.OverlayWidget)
+NotesOverlay.ATTRS{
+    desc='Render map notes.',
+    viewscreens='dwarfmode',
+    default_enabled=true,
+    overlay_onupdate_max_freq_seconds=30,
+}
+
+local waypoints = df.global.plotinfo.waypoints
+local map_points = df.global.plotinfo.waypoints.points
+
+function NotesOverlay:init()
+    self.notes = {}
+    self.note_manager = nil
+    self.last_click_pos = {}
+    self:reloadVisibleNotes()
+end
+
+function NotesOverlay:overlay_onupdate()
+    self:reloadVisibleNotes()
+end
+
+function NotesOverlay:overlay_trigger(args)
+    return self:showNoteManager()
+end
+
+function NotesOverlay:onInput(keys)
+    if keys._MOUSE_L then
+        local top_most_screen = dfhack.gui.getDFViewscreen(true)
+        if dfhack.gui.matchFocusString('dwarfmode/Default', top_most_screen) then
+            local pos = dfhack.gui.getMousePos()
+
+            local note = self:clickedNote(pos)
+            if note ~= nil then
+                self:showNoteManager(note)
+            end
+        end
+    end
+end
+
+function NotesOverlay:clickedNote(click_pos)
+    local pos_curr_note = same_xyz(self.last_click_pos, click_pos)
```

Comment:
should protect against `click_pos` being `nil`, which I believe can happen if the player clicks on the screen, but outside of the map area